drivers/include/extend/gpio.h

```diff
@@ -0,0 +1,491 @@
+ */
+typedef struct gpio_ext_driver {
+    gpio_ext_init_t init;                 /**< callback for @ref gpio_init */
+#if MODULE_PERIPH_GPIO_IRQ || DOXYGEN
+    gpio_ext_init_int_t init_int;         /**< callback for @ref gpio_init_int */
+    gpio_ext_irq_enable_t irq_enable;     /**< callback for @ref gpio_irq_enable */
+    gpio_ext_irq_disable_t irq_disable;   /**< callback for @ref gpio_irq_disable */
+#endif /* MODULE_PERIPH_GPIO_IRQ || DOXYGEN */
+    gpio_ext_read_t read;                 /**< callback for @ref gpio_read */
+    gpio_ext_set_t set;                   /**< callback for @ref gpio_set */
+    gpio_ext_clear_t clear;               /**< callback for @ref gpio_clear */
+    gpio_ext_toggle_t toggle;             /**< callback for @ref gpio_toggle */
+    gpio_ext_write_t write;               /**< callback for @ref gpio_write */
+} gpio_ext_driver_t;
+
+/**
+ * @brief   GPIO extension device type
+ */
+typedef struct gpio_ext_dev {
+    const gpio_ext_driver_t *driver;    /**< pointer to extension device driver */
+    void *dev;                          /**< pointer to extension device descriptor */
+} gpio_ext_dev_t;
+
+#include "gpio_ext_conf.h"
+
+/**
+ * @name    GPIO API functions realized by the GPIO extension interface.
+ * @{
+ */
+
+/**
+ * @brief   Implementation by the GPIO extension interface, see @ref gpio_init
+ */
+static inline int gpio_init(gpio_t pin, gpio_mode_t mode)
+{
+    /*
+     * Check that the GPIO_EXT_BIT is inside the gpio_t size.
+     */
+    _Static_assert(GPIO_EXT_BIT < (sizeof(gpio_t) * 8),
+                   "GPIO_EXT_BIT has to be less than sizeof(gpio_t)*8");
+    /*
+     * Check that the gpio_t value has at least one bit left for the device
+     * number to address GPIO extension devices.
+     */
+    _Static_assert(GPIO_EXT_PIN_BITS < GPIO_EXT_BIT,
+                   "GPIO_EXT_PIN_BITS has to be less than GPIO_EXT_BIT");
+    /*
+     * Check that the number of bits used for the device number is large enough
+     * to address all extension devices in the device list.
+     */
+    _Static_assert((1 << (GPIO_EXT_BIT - GPIO_EXT_PIN_BITS)) >= GPIO_EXT_DEV_NUMOF,
+                   "Number of bits for GPIO extension device number is too small");
+
+    if (GPIO_EXT_IS(pin)) {
+       assert(GPIO_EXT_PIN_DEV(pin) < ARRAY_SIZE(gpio_ext));
+       return GPIO_EXT_DEV(pin).driver->init(GPIO_EXT_DEV(pin).dev,
+                                             GPIO_EXT_PIN_NUM(pin), mode);
+    }
+#if MODULE_PERIPH_GPIO
+    return gpio_cpu_init(pin, mode);
+#endif
+    return -1;
+}
+
+#if MODULE_PERIPH_GPIO_IRQ || DOXYGEN
+
+/**
+ * @brief   Implementation by the GPIO extension interface, see @ref gpio_init_int
+ */
+static inline int gpio_init_int(gpio_t pin, gpio_mode_t mode,
+                                gpio_flank_t flank, gpio_cb_t cb, void *arg)
+{
+    /*
+     * Check that the GPIO_EXT_BIT is inside the gpio_t size.
+     */
+    _Static_assert(GPIO_EXT_BIT < (sizeof(gpio_t) * 8),
```

Comment:
This only works with C11 compilers. In `core/include/assert.h` a macro `static_assert()` is defined that wraps `_Static_assert()` on C11 compilers and later, but provides a fall back implementation for older ones. I'd prefer this version for compatibility reasons

Reply:
While `_Static_assert` didn't throw any compiler error, I had compilation problems when I was trying to use `static_assert` from `core/include/assert.h`.

Reply:
Hmm, then the one provided in `assert.h` should be fixed. But I agree that this is out of scope of this PR. (@maribu: Note to myself: Check this.)

Reply:
The problem was that `static_assert` can only happen once within a scope, but there are three tests. Although the conditions could be ANDed, it would be difficult to identify which one fails.

A solution would be to embed each `static_assert` into curly braces. But I would tend to extend the `static_assert` macro in `core/include/assert.h` by curly braces.